_posts/2021-05-27-TC-Guard.md

```diff
@@ -0,0 +1,162 @@
+This is also used to expose the Graylog dashboard to the public.
+TC Guard and Graylog are connected so that the logs from TC Guard can be sent to Graylog.
+The data collected by the Android application is stored in the PostgreSQL database via the TC Guard API.
+When the application needs this data, it is retrieved again via the TC Guard API.
+The last block connected to the TC Guard API is Prometheus.
+Prometheus is used to observe the API.
+The resulting data is then displayed in the Grafana dashboard.
+
+### Argo CD
+Argo CD is a Continuous Delivery GitOps tool for Kubernetes.
+Within the internship assignment, this is used to install helm charts on the Kubernetes cluster.
+If a change is made on the Azure DevOps repository, Argo CD will detect this and implement the changes on the Kubernetes cluster via a rolling update.
+
+It was quite satisfying to observe the changes being applied.
+Argo CD made managing our applications much easier.
+For example when we updated Graylog the only thing we had to do was change the version number in a .yaml file and Argo CD took care of everything else.
+
+<img src = "/img/2021-05-27-TC-Guard/argocd.png" class = "image fit">
+
+### Ingress
+Ingress provides routing rules to manage remote users' access to services in a Kubernetes cluster. With Ingress, simple rules can be set for routing traffic without exposing every service on the endpoint.
+While setting up Ingress, a few issues surfaced due to our inexperience with Kubernetes.
+Because the online documentation of Ingress nginx was rather limited, setting up Ingress took a fair amount of time.
+Afterwards, Ingress nginx was replaced by Contour.
+By this time we had more experience with Kubernetes.
+Because of this, the switch to Contour went smoothly. 
+
+### Graylog
+The centralized logging system aims to obtain the logs of the API and the Android application in a central place so that it would be easier to detect errors.
+To obtain and view the logs in a central place, Graylog is used.
+We chose Graylog because it is much easier to use than Kibana.
+We knew that Kibana has more features than Graylog, but for this project it was not required.
+
+<img src = "/img/2021-05-27-TC-Guard/graylog.png" class = "image fit">
+
+### Grafana
+Prometheus was used to observe the API.
+Prometheus collects metrics from targets by scraping given HTTP endpoints.
+Actuator is a Spring Boot plugin that is used to manage and monitor your application by using HTTP endpoints. 
+This makes it possible for Prometheus to gather metrics and monitor the health of the application. 
+Grafana then uses Prometheus as a source to display the metrics on a custom dashboard. 
+
+We were unsure how to add a target to Prometheus while it was running on Kubernetes. 
+Because we were not sure how to modify the config file, a lot of time was lost.
+To solve this problem we had to create a secret that contains the target.
+This secret should then be added to the prometheus.yaml file.
```

Comment:
Inline code markup makes it easier for the reader to spot inline code but also (config) file references for example.
```suggestion
This secret should then be added to the `prometheus.yaml` file.
```